service/storage_service.cc

```diff
@@ -2244,10 +2246,27 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
             auto& db = ss._db.local();
             for (auto& ks_name : keyspaces) {
                 auto rs = db.find_keyspace(ks_name).get_replication_strategy_ptr();
+                if (rs->is_per_table()) {
```

Comment:
Yes. A preliminary loop clones shard 0's token_metadata on all other shards
and then it proceeds to the next step which is generating the modified e_r_m using `get_erm_factory().create_effective_replication_map`

Shard 0 gets the result from this call
and the other shards are expected to find it in the erm factory cache.

Reply:
Shouldn't the tablets map be part of the e_r_m?

Reply:
> i.e. the per-table erms could be stored in the per-keyspace erm.

It seems cleaner to store them directly in the `table` object because that's where they really belong (they're per table). We avoid double lookup this way, you only lookup the table and get the erm from there, instead of first the keyspace and then the table in that keyspace.

Extending the per-keyspace erm with per-table lookup capabilities seemed like unnecessary complication in the end.

Reply:
> Shard 0 gets the result from this call
> and the other shards are expected to find it in the erm factory cache.

If we cloned directly from here, would it mean we don't need the cache?

Reply:
> Keeping the tablets map separately means we'll need to come up with an equivalent mechanism just for tablets, while we can reuse the existing mechanism easily.

Tablet maps are kept in token_metadata, not separately. Do you mean per-table erms? What kind of tracking are we talking about?

Reply:
> > If we cloned directly from here, would it mean we don't need the cache?
> 
> I think we clone in the cache also when creating the initial version from

Right.

Reply:
> IIUC, the tablet maps contain the full replication information. Does it even have a notion of a primary owner of a tablet?

It doesn't.

> If so, I think they should live as a member in the e_r_m, rather in the token_metadata object.

erm = rs_instance + token_metadata

We cannot make erm the primary owner of tablet_map, because erm is computed, based on other structures with longer lifetime. We would have to manage tablet_metadata like we manage token_metadata, and make replication_strategy interfaces accept both. It's just simpler to make tablet_metadata a member of token_metadata, and say that token_metadata is all metadata related to topology which is not computed based on options, but fundamental (read from system tables). It already is, it stores rack and dc assignment, not just tokens. So it can also store tablet metadata.

Reply:
> > IIUC, the tablet maps contain the full replication information. Does it even have a notion of a primary owner of a tablet?

> It doesn't.

I think I misunderstood the question. The owner of tablet_map is token_metadata (it's an embedded member, not a pointer). tablet_map is versioned together with token_metadata. New version of table_map is created by creating a new version of token_metadata.

Reply:
So it's looked up in token_metadata or one of the similar-lifetime objects, not table?

Reply:
Yes, erm looks up in tablet_metadata. This could be optimized by storing a pointer in the erm.